### PR TITLE
Telegram bot

### PR DESCRIPTION
**Pull request by alessioGalatolo**
Refactoring data crawler to ease the integration of the bot.
Now the interaction with the database is segregated into its own module.
Data crawler is now a class.
External tools package is not a subpackage of the flask app anymore to favor testing.
Add data_crawler pytests.
Add email sending when a new API is needed.

### CODE REVIEW

**Comment by mortorit** [comment]
private (whitelisted access) Telegram bot for making interaction with the server easier.
Current functionalities:

- set api key
- get server log
- retrieve info about crawling process 

Added unit testing for whitelisting and API key checking and setting.

File: mooncaker/flask_app/__init__.py [hunk 1]
```diff
@@ -70,5 +68,3 @@ def get_api_key():
 bot_process = Process(target=bot.start_bot)
 bot_process.start()
 logging.info("mooncaker: starting telegram bot")
-
-from flask_app import routes
```
Comment (alessioGalatolo) [inline mooncaker/flask_app/__init__.py]: This breaks our entire website 😢

File: mooncaker/external_tools/mooncaker_bot.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            else:
+                context.bot.send_document(chat_id=update.effective_chat.id, document=log,
+                                          filename="full_log.txt")
+
+    def set_api_key_req(self, update: Update, context: CallbackContext) -> int:
+        update.message.reply_text(
+            f"Please gimme all your new API key, I want it so bad..or use /cancel to make me stop waiting..")
+        return self.WAITING_API
+
+    def set_new_api(self, update: Update, context: CallbackContext, set_api):
```
Comment (alessioGalatolo) [inline mooncaker/external_tools/mooncaker_bot.py]: set_api is saved inside object, no need to pass it

**Review by mortorit — APPROVED** [review]

File: mooncaker/external_tools/mooncaker_bot.py (hunk 2 above)
Comment (mortorit) [reply mooncaker/external_tools/mooncaker_bot.py]: yup, I forgot to use the object variable when converting the bot to a class 😢

**Comment by alessioGalatolo** [comment]
@mortorit can you add some short docs to the telegram bot class?

File: mooncaker/external_tools/mooncaker_bot.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+    def __init__(self, token, set_api, log_url, whitelist, db_url="mongodb://datacaker:27017"):
+        self.db = Database(db_url)
+        self.token = token
+        self.set_api = set_api
+        self.log_url = log_url
+        self.whitelist = whitelist
+        self.WAITING_API = 0
+
+    def authorize_and_dispatch(self, update: Update, context: CallbackContext, dispatcher, whitelist):
```
Comment (alessioGalatolo) [inline mooncaker/external_tools/mooncaker_bot.py]: Whitelist is already given on object init. Remember to change pytest accordingly

File: tests/test_telegram_bot.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+import pytest
+from mooncaker.external_tools.mooncaker_bot import MooncakerBot
+from multiprocessing import Queue
+from telegram import Update, User, Message, Chat
+
+
+@pytest.fixture()
+def bot():
+    return MooncakerBot('not_a_token', None, 'not_a_path', 'not_a_whitelist',None)
```
Comment (alessioGalatolo) [inline tests/test_telegram_bot.py]: No whitelist is given, add one in order to properly test authorized users

**Review by mortorit — APPROVED** [review]

**Pull request merged**